compiler-core/src/type_/expression.rs

```diff
@@ -3507,6 +3513,37 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
             self.minimum_required_version = minimum_required_version;
         }
     }
+
+    /// When targeting JavaScript, adds a warning if the given Int value is outside the range of
+    /// safe integers as defined by Number.MIN_SAFE_INTEGER and Number.MAX_SAFE_INTEGER.
+    ///
+    fn check_javascript_int_safety(&mut self, value: &EcoString, location: SrcSpan) {
+        if self.environment.target != Target::JavaScript {
+            return;
+        }
+
+        // Determine the radix based on the prefix
+        let (radix, bytes) = if let Some(value) = value.strip_prefix("0x") {
+            (16, value.as_bytes())
+        } else if let Some(value) = value.strip_prefix("0o") {
+            (8, value.as_bytes())
+        } else if let Some(value) = value.strip_prefix("0b") {
+            (2, value.as_bytes())
+        } else {
+            (10, value.as_bytes())
+        };
+
+        // Parse value as a bigint and check against JavaScript's min/max safe integer
+        if let Some(i) = num_bigint::BigInt::parse_bytes(bytes, radix) {
+            let js_min_safe_integer = -9007199254740991i64;
+            let js_max_safe_integer = 9007199254740991i64;
+
+            if i < js_min_safe_integer.into() || i > js_max_safe_integer.into() {
+                self.problems
+                    .warning(Warning::JavaScriptIntUnsafe { location });
+            }
+        }
+    }
```

Comment:
Seeing as we want to use the int values in a few places should we do this string-to-int during parsing? To avoid each step of the compiler parsing it over and over

Reply:
The most recent commit to this branch attempts this. I'm not 100% sure it's what you were meaning, but it might be!

- `Token::Int` gets an `int_value: Option<BigInt>` field, which flows through to `UntypedExpr::Int`, `TypedExpr::Int`, `Pattern::Int`, and `Constant::Int`.
- This field is then used for the new warning.

Reply:
Very nice! Why is it optional? Shouldn't it always be present?

Reply:
Yes you're right, I've made it always present and fixed handling of leading underscores on parsing as bigint.

There are now three `.expect("int value to parse as bigint")` calls, but I believe these to be safe due to previous validation done during parsing.